Assetlib: Fix long plugin names breaking the UI

GrammAcc opened the pull request:
Edit:
I think I have this fixed now.

Assets should have their titles truncated with an ellipsis to within the column width now, and I added a tooltip with the full asset title to the ones that get truncated, so that the user can still see the full title without opening the asset page.

I also changed the pagination from having 10 fixed page buttons to using the `EDSCALE` to dynamically choose a number of nav buttons to display so that it won't cause the horizontal overflow issue on large scales but will still show more page buttons on smaller scales.

It works on my machine at all scales up to 200%. At custom scales above 200% it starts to overflow again, but this happens regardless of the title and nav buttons. I tested this by truncating all titles to three characters and removing the nav buttons completely, and it still overflows at over 200% scale, so I think that is a separate issue and probably affects other parts of the engine UI as well.

I haven't written a unit test for this yet, but I can add one if needed. It will likely be next Saturday before I am able to get to it though.

Here are the updated screenshots:

<details>
<summary>
75% scale:
</summary>

![75percent](https://github.com/godotengine/godot/assets/35282898/75d5dc90-f9a2-4e39-9809-a3f606439d99)

</details>

<details>
<summary>
100% scale:
</summary>

![100percent](https://github.com/godotengine/godot/assets/35282898/03225be2-c910-42b0-bdea-c278433fe56d)

</details>

<details>
<summary>
125% scale:
</summary>

![125percent_top](https://github.com/godotengine/godot/assets/35282898/04d7c7a6-9a0f-415a-823c-0c462fe554e6)

![125percent_longtitle](https://github.com/godotengine/godot/assets/35282898/f3d819ae-c62f-4140-9406-51921f4a5ad0)

</details>

<details>
<summary>
150% scale:
</summary>

![150percent_top](https://github.com/godotengine/godot/assets/35282898/f3b390a5-70a5-4dd3-b552-3968cdabaee0)

![150percent_longtitle](https://github.com/godotengine/godot/assets/35282898/c2296b19-28ec-42f2-b629-86f02893d171)

</details>

<details>
<summary>
175% scale:
</summary>

![175percent_top](https://github.com/godotengine/godot/assets/35282898/c4a96993-b0cf-4a35-be6f-b074aecfda71)

![175percent_longtitle](https://github.com/godotengine/godot/assets/35282898/85023530-bdd1-4532-b6b0-c829276f3a3a)

</details>

<details>
<summary>
200% scale:
</summary>

![200percent_top](https://github.com/godotengine/godot/assets/35282898/e7a599ec-d391-4569-810e-106542824a80)

![200percent_longtitle](https://github.com/godotengine/godot/assets/35282898/aed0864a-3c8f-409c-b0a5-9791a9f385bd)

</details>

Original PR:
I removed the line that disables horizontal scrolling in the AssetLib plugin. This fixes the screen overflow issue on my machine, but I am not sure if this is an appropriate solution since there was probably a reason why scroll was disabled in the first place, and I was unable to reproduce the column alignment issues seen in the screenshots posted in #80507 on either the current master or 4.1.1 stable, so I think there may be something platform specific going on here.

My system:
- Linux version 6.4.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) Thu, 03 Aug 2023 16:02:01 +0000

Please advise if enabling horizontal scroll in the AssetLib is a problem, and I will look for another solution.

Additionally, I wanted to write a unit test for this that would basically just check that the editor window size didn't exceed the size of the screen, but I wasn't sure which classes to use for that, and I wanted to get feedback on this solution early, so I'm submitting this now, but if someone can point me in the right direction for where to look for methods related to overall editor window size and screen size, that would be much appreciated!

Here's some screenshots of the AssetLib on my local machine with this fix in place at different editor scales:
<details>
<summary>100%</summary>
<br/>

![assetlib_100percent](https://github.com/godotengine/godot/assets/35282898/63377e71-9c1b-45cd-ba21-c9211228a4ed)

</details>
<details>
<summary>125%</summary>
<br/>

![assetlib_125percent](https://github.com/godotengine/godot/assets/35282898/980aefee-e0c4-48be-82c6-3a0d815baad4)

</details>
<details>
<summary>150%</summary>
<br/>

![assetlib_150percent](https://github.com/godotengine/godot/assets/35282898/e0b87142-644a-4549-8689-18ff298f37a6)
</details>

<details>
<summary>175%</summary>
<br/>

![assetlib_175percent](https://github.com/godotengine/godot/assets/35282898/d997779c-b6b4-4764-8194-3f0204f8f1c2)

</details>
<details>
<summary>200%</summary>
<br/>

![assetlib_200percent](https://github.com/godotengine/godot/assets/35282898/995e27de-2013-4493-adf8-723d06cc4bff)

</details>

* *Bugsquad edit, fixes: #80507*

GrammAcc commented:
Looks like the screenshots aren't displaying correctly in the PR. I don't have time to mess with them right now. I'll fix them later today. :)

AThousandShips commented:
For screenshots to display you need to put spaces around it, I have fixed it, but keep that in mind for the future

akien-mga commented:
Thanks for the contribution!

I don't think horizontal scroll is a good solution for this issue, especially if it changes unpredictability on each page due to the names of assets.

I think my suggestion to trim (or alternatively wrap) long names is better, so we keep the column width constant and avoid the bad UX which is having to scroll both horizontally and vertically.

GrammAcc commented:
> Thanks for the contribution!
> 
> I don't think horizontal scroll is a good solution for this issue, especially if it changes unpredictability on each page due to the names of assets.
> 
> I think my suggestion to trim (or alternatively wrap) long names is better, so we keep the column width constant and avoid the bad UX which is having to scroll both horizontally and vertically.

I thought that might be the case. Thank you for confirming!

I apologize for not including this in the PR description. While I was working on this earlier, the first thing I tried was simply truncating the asset title to three characters to see what would happen, and it didn't fix the issue, and there were no other fields that were especially wide, so I started looking at the container setup code for size/margin calculations, and when I saw that horizontal scroll was disabled, I went with the simplest solution to the problem as I saw it.

I will need to investigate this further to come up with a proper solution.

FWIW, I was not able to find anything in the code that stood out to me as setting a fixed width or that would cause the UI to expand further than its content, so I'm guessing it has something to do with the width of fields other than the asset title, or that there is something going on with the column width calculations, which should be evenly spread between the number of columns determined as `get_size().x / (450 * EDSCALE)`, so I would think that the columns would not expand past their content, but I may be misunderstanding the `SIZE_EXPAND_FILL` scaling flag. I will look into this as well. :)

I will try to get an update pushed as soon as I can, but I probably won't be able to get anything done until next Saturday since I work six days/week.

Thanks!

GrammAcc commented:
Upon further investigation, it looks like the nav buttons for the paginated results are what is actually causing the overflow on my system, but I still have not been able to reproduce the uneven column width behavior shown in the screenshots in #80507 on my machine, so now I'm not sure if a fix on my machine would constitute a fix of the issue as reported, and I don't have any other system on which I can test this.

I will investigate further and see if I can get to the root of the issue by next weekend.

GrammAcc commented:
Sorry for blowing up the conversation.

I was able to reproduce the differently sized columns on my local machine by searching for "debug" to find an asset with a sufficiently long name for my DPI settings, so I think just truncating the content length to the calculated column width to keep all columns evenly spaced should fix the issue with a long asset title causing column expansion problems, and I will also look into the scaling/expansion of the nav buttons as well since they cause a similar issue on my machine. I will get to work on these fixes as I am able.

I apologize for all of the confusion. I should have put more time up front into reproducing the issue exactly as it was reported instead of focusing on the screen overflow.

AThousandShips commented:
Please use rebasing for updating your branch instead of merging, see [here](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html)

GrammAcc commented:
I'm very sorry for the mess that the git history has become. This is my first time contributing to a large project, so I was unprepared to deal with the rebase/merge workflow, and after wrangling with my local fixup commits, I decided to just reset to the state I started in, but this appears to have auto-closed this PR.

I will get my fork cleaned up and start working with the correct workflow. I can open a new PR or reopen this one once I have a new commit ready for review. I am still working on a fix for this bug though.

AThousandShips commented:
Good luck! If this branch gets cleaned up this can be reopened if you push to it

GrammAcc commented:
I was hoping to have something pushed today, but I don't have any more time to work on this right now. I will keep working on this and try to get something pushed next weekend.

GrammAcc commented:
@AThousandShips I pushed a commit to my branch that I believe will fix this issue. Can you reopen this PR?

Thanks!

GrammAcc commented:
It looks like there is actually still an issue with some overflow at 200% scale. I think I can fix this by tweaking the scaling numbers I'm using. I will amend my commit once I have this sorted out.

GrammAcc commented:
Got the slight overflow at 200% scale fixed. Should be ready for review now.

Calinou commented:
Out of curiosity, why not use the built-in ellipsis feature found in Label?

Calinou reviewed (approved):
Tested locally (rebased on top of `master` bc88dca176cef6b8943ded58c68a60f7c127baea), it works as expected.

Before | After
-|-
![Screenshot_20230830_184713](https://github.com/godotengine/godot/assets/180032/fea2bae5-8b60-4584-a082-9d2647fd4991) | ![Screenshot_20230830_184845](https://github.com/godotengine/godot/assets/180032/cb63515f-3ed4-45e1-a021-5dc350928ccf)

GrammAcc commented:
> Out of curiosity, why not use the built-in ellipsis feature found in Label? This was added in 4.0 though, so the fix can't be backported to 3.x if we decide to use that.

Sorry, I was not aware of this feature, but I can change it if desired.

If it is part of `Label`, I think I would have to change the displayed `title` text to use `Label` instead of the `LinkButton` directly and then just draw the `LinkButton` over the `Label` to keep the functionality. I'm not sure what functionality of `LinkButton` is actually needed here, so I didn't want to remove it naively.

On the other hand, if I did it that way, I would personally prefer to just use word wrapping in the `Label` instead of truncating the text since that would be consistent with the UI on the web version of the Asset Library.

What do you think?

Calinou commented:
> If it is part of Label, I think I would have to change the displayed title text to use Label instead of the LinkButton directly and then just draw the LinkButton over the Label to keep the functionality. I'm not sure what functionality of LinkButton is actually needed here, so I didn't want to remove it naively.

Unfortunately, LinkButton lacks text ellipsis functionality. It's present in Button though, so it could likely be ported in a separate PR.

I think we can merge this PR as-is and look at improving the fix later.

GrammAcc commented:
Sounds good. Thanks for confirming!

AThousandShips commented on editor/plugins/asset_library_editor_plugin.h:
```diff
@@ -305,6 +307,8 @@ class EditorAssetLibrary : public PanelContainer {
 
 	void _install_external_asset(String p_zip_path, String p_title);
 
+	int _asset_items_column_width;
```
```suggestion
	int asset_items_column_width;
```
While having special prefixes for member variables would be desirable we don't prefix with `_` normally

GrammAcc replied:
I amended the commit with this change introduced. This is good to review now.

GrammAcc commented:
I need to experiment a bit with amending commits. It looks like the soft-reset > force push > commit amended changes > push again process resulted in the PR being auto-closed again.

AThousandShips commented:
It auto closes if you make a push that is equivalent to the branch it wants to merge into, as it no longer is a change

YuriSizov commented on editor/plugins/asset_library_editor_plugin.cpp:
```diff
@@ -63,6 +63,21 @@ void EditorAssetLibraryItem::configure(const String &p_title, int p_asset_id, co
 }
 
+void EditorAssetLibraryItem::clamp_width(int p_max_width) {
+	int text_pixel_width = title->get_theme_default_font().ptr()->get_string_size(title->get_text()).x * EDSCALE;
```
Why theme default font? You should use the font that the link button actually uses.

In this case that would be `title->get_theme_font(SNAME("font"))`. But we should avoid string access. So I suggest you add a new public method to `LinkButton`, say, `LinkButton::get_button_font()` which returns its `theme_cache.font`, and use it here.

GrammAcc replied:
Great catch! I couldn't figure out how to properly access the theme font, and I completely spaced on the fact that the default font won't always be the font used in the UI. :facepalm:

I added the getter to the `LinkButton` in the update I just pushed. I was initially going to add a setter, but I wasn't sure exactly how the `theme_cache` integrated with the rest of the theming system, so I wasn't sure if setting the font on the cache was appropriate, so I did not implement a setter for it. Please let me know if you want me to add a setter as well. Thanks!

YuriSizov replied:
You shouldn't add a setter. You should only read from the cache. It is set automatically when appropriate by the control itself.

YuriSizov commented on editor/plugins/asset_library_editor_plugin.cpp:
```diff
@@ -63,6 +63,21 @@ void EditorAssetLibraryItem::configure(const String &p_title, int p_asset_id, co
+		String full_text = title->get_text();
+		title->set_tooltip_text(full_text);
+
+		int text_len = full_text.length();
```
No need to extract this into a variable, it's clear what `full_text.length()` is and it's short enough to be used directly in the next line.

GrammAcc replied:
Fixed. Thanks!

YuriSizov commented on editor/plugins/asset_library_editor_plugin.cpp:
```diff
@@ -63,6 +63,21 @@ void EditorAssetLibraryItem::configure(const String &p_title, int p_asset_id, co
+	if (text_pixel_width > p_max_width) {
+		// Truncate title text to within the current column width.
+		String full_text = title->get_text();
+		title->set_tooltip_text(full_text);
```
I think we can just always set the tooptip, it never hurts.

GrammAcc replied:
I moved these two lines above the conditional. Thanks!

YuriSizov reviewed (changes requested):
Should be fine after a couple of changes, but this should really be redone using TextServer at some point. Might be worth adding a TODO comment next to the new method suggesting as much.

GrammAcc commented:
Thank you for the detailed review! I will get these changes implemented as soon as I can, but I have a really busy weekend ahead of me, so it may be about a week before I get this resubmitted. I should have this done by Saturday of next week at the latest.

GrammAcc committed:
AssetLib: Fix long plugin names breaking the UI

The UI was extending past the screen width when loading a page diplaying
a plugin with an especially long title in the asset store plugin.

I implemented a new `EditorAssetLibraryItem::clamp_width` method that
checks that the title text is not longer than the column width minus
some padding and truncates it if it is.

I also noticed that the nav buttons for paginated results were causing the UI to extend past
the screen width on higher editor scales since they were hardcoded to
show ten page buttons if there were enough results. I modified the
pagination slightly to display a dynamic number of nav buttons based on
the editor scale in order to fix this other cause of the same problem.

I had to use the font of the `title`, which is a `LinkButton` in order
to determine the text width, so I added a public getter `get_button_font` to the `LinkButton` class.

GrammAcc commented on editor/plugins/asset_library_editor_plugin.h:
```diff
@@ -305,6 +307,8 @@ class EditorAssetLibrary : public PanelContainer {
 
 	void _install_external_asset(String p_zip_path, String p_title);
 
+	int asset_items_column_width;
```
This is fixed in the update I pushed today. :)

GrammAcc commented:
@YuriSizov I was able to carve out some time for this today. :)

I implemented all of the requested changes and pushed the fixed commit. I also added a more detailed commit message this time.

This should be ready for review again. Thanks!

GrammAcc commented:
I didn't realize I could just re-request a review. Sorry! It's been a long day today.

YuriSizov reviewed (approved):
LGTM, good work!

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:

YuriSizov commented:
Cherry-picked for 4.1.3.
